# Práctica Nro. 4

[![Abrir en Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/manuxch/calculo_avanzado/blob/main/04-errores/code/practica_04.ipynb)

En esta práctica utilizamos solamente las funciones `floor()` y `log10()` del módulo math de Python. La documentación de las funciones definidas en este módulo se pueden ver [aquí](https://docs.python.org/es/3/library/math.html).

In [ ]:
from math import floor, log10

## Ejercicio 1

##### Al considerar que usamos solo cinco cifras significativas, debemos utilizar el redondeo en caso de que nuestro número original tenga más de cinco cifras. Este es el caso, por ejemplo, del número `0.000924138`. Veamos cómo hacer paso a paso el redondeo definido por
$$ \mathcal{R}(x) = [b^n \times 0.m + 0.5] \times 10^{e-b} $$
donde para este ejercicio, $n = 5$ y $x = 0.924138 \times 10^{-3}$, con lo que $e = -3$, y $[x]$ denota la parte entera de $x$. Entonces:
\begin{align}
\mathcal{R}(x) &= \mathcal{R}(0.924138 \times 10^{-3}) \\
&= [0.924138 \times 10^{5} + 0.5] \times 10^{-8} \\
&= [92413.8 + 0.5] \times 10^{-8} \\
&= [92414.3] \times 10^{-8} \\
&= 92414 \times 10^{-8} \\
&= 0.00092414
\end{align}

Una vez redondeado el número a la cantidad de cifras significativas las expresamos en punto flotante, lo que equivale a expresarlo en notación científica:
- `84.175` = $8.4175 \times 10^1$
- `-528.685` = $-5.2869 \times 10^2$
- `0.000924138` = $9.2414 \times 10^{-4}$
- `-362005` = $-3.6201 \times 10^5$

Python permite representar números en formato de punto flotante, mediante la especificación de la cantidad de cifras después del punto decimal a mostrar en notación científica. Como en esta notación se utiliza solo un dígito para la parte entera, si queremos mantener solo cinco cifras significativas debemos especificar entonces cuatro dígitos para la parte decimal. El siguiente código genera una lista con los números del ejercicio, y luego un bucle `for` itera sobre esta lista imprimiendo el número original y su representación en notación científica con cinco cifras significativas (uno para la parte entera y cuatro para la decimal):

In [2]:
numeros = [84.175, -528.685, 0.000924138, -362005]
for numero in numeros:
    print(f"{numero} -> {numero:1.4E}")

84.175 -> 8.4175E+01
-528.685 -> -5.2868E+02
0.000924138 -> 9.2414E-04
-362005 -> -3.6200E+05


Podemos ver que esta representación no coincide con el cálculo que realizamos previamente. Esto se debe a que el redondeo se realiza en forma diferente al que hemos propuesto, y a la representación aproximada de los números en la computadora. Para ver esto en detalle, generamos una función `R(x, n)` que redondea el número `n` que le pasamos como primer argumento a la cantidad de cifras significativas `n` que le pasamos como segundo argumento:

In [3]:
def R(x, n):
    e = int(floor(log10(abs(x)))) + 1
    m = x * 10**(-(e))
    r1 = 10**n * m + 0.5
    r = int(r1) * 10**(e - n)
    print(f'{e=}, {m=}, {r1=}, [r1]={int(r1)}, {r=}')
    return r

La instrucción `print()` nos muestra los valores intermedios y final del cálculo que realiza. Si ahora ejecutamos el bucle sobre la lista de números vemos:

In [4]:
for x in numeros:
    if x >= 0:
        rx = R(x, 5)
    else:
        rx = -R(-x, 5)
    print(f"x = {x}, R(x) = {rx}, Python: {x:1.4E}")

e=2, m=0.84175, r1=84175.5, [r1]=84175, r=84.175
x = 84.175, R(x) = 84.175, Python: 8.4175E+01
e=3, m=0.528685, r1=52868.99999999999, [r1]=52868, r=528.6800000000001
x = -528.685, R(x) = -528.6800000000001, Python: -5.2868E+02
e=-3, m=0.924138, r1=92414.3, [r1]=92414, r=0.00092414
x = 0.000924138, R(x) = 0.00092414, Python: 9.2414E-04
e=6, m=0.36200499999999997, r1=36201.0, [r1]=36201, r=362010
x = -362005, R(x) = -362010, Python: -3.6200E+05


Vemos para el segundo número que `r1=52868.99999999999`, que es la representación de máquina del número $52869$, por lo que al tomar la parte entera ($[x]$) nos queda `52868` en vez de `52869`. Si bien la diferencia es pequeña (del orden del 0.002%), realizando muchos cálculos los errores acumulados pueden ser importantes.

## Ejercicio 2

Resolveremos la ecuación $x^2 - 300 x + 1 = 0$ de dos maneras, manteniendo primero cuatro cifras significativas y luego dos.

Utilizando la fórmula cuadrática:
$$ x_{1,2} = \frac{1}{2a} \left( -b \pm \sqrt{b^2 - 4 a c} \right) $$
con $a = 1$, $b = -30$ y $c = 1$, tenemos para la raíz cuadrada:
$$ \sqrt{(-30)^2 - 4} = \sqrt{900 - 4} = \sqrt{896} = 29.922 = 29.93 $$ 
Entonces, calculando $x_1$ y $x_2$ con cuatro cifras significativas tenemos:
$$ x_1 = \frac{1}{2} (30 + 29.93) = \frac{1}{2} 59.93 = 29.965 = 29.97 $$
y 
$$ x_2 = \frac{1}{2} (30 - 29.93) = \frac{1}{2} 0.07 = 0.035 $$
Es importante notar que al calcular $x_2$ a partir de números con cuatro cifras significativas, obtenemos un resultado que tiene solo dos cifras significativas. Hemos perdido precisión en el cálculo.

Como método alternativo de solución para $x_2$, podemos usar:
$$ x_{1,2} = \frac{1}{2a} \left( -b + \sqrt{b^2 - 4 a c} \right), \quad x_2 = \frac{c}{a x_1} $$
La raíz $x_1$ (calculada con la fórmula que suma números de magnitud similar) es $29.97$ como antes, pero ahora para $x_2$ obtenemos:
$$ x_2 = \frac{c}{a x_1} = \frac{1}{29.97} = 0.0333667 = 0.03337 $$
que tiene **cuatro** cifras significativas.

Si ahora repetimos el cálculo con 2 cifras significativas, el resultado de la raíz cuadrada es
$$ \sqrt{(-30)^2 - 4} = \sqrt{900 - 4} = \sqrt{900} = 30 $$
con **dos** cifras significativas (el redondeo de 900 - 4 = 896, expresado con dos cifras significativas es 900). Entonces, para $x_1$ tenemos
$$ x_1 = \frac{1}{2} (30 + 30) = \frac{1}{2} 60 = 30 $$
y 
$$ x_2 = \frac{1}{2} (30 - 30) = 0 $$
Del mismo modo que para el caso anterior, podemos obtener una mejor estimación para $x_2$ si hacemos
$$ x_2 = \frac{c}{a x_1} = \frac{1}{30} = 0.033333 = 0.033 $$
con **dos** cifras significativas.

El propósito de este ejercicio no es mostrar que realizar el cálculo con menos cifras significativas genera resultados más pobres (lo cual es esperable, aunque no siempre es el caso). El punto es mostrar, en términos de números simples, lo que sucede en principio independientemente de la cantidad de cifras significativas usadas en el cálculo. El uso de la fórmula cuadrática para obtener ambas raíces involucra la diferencia de números de similar magnitud, lo que hace perder precisión en el cálculo. Esto es fácilmente reconocible haciendo las cuentas *a mano*, pero no es tan simple de ver en un cálculo computacional donde solo se muestran los resultados finales. Esto explica la necesidad de realizar programas que tiendan a evitar la pérdida de precisión en cálculos aritméticos (lo que en este caso es posible al evaluar $x_2$ en forma alternativa).

## Ejercicio 3

Si $x_2$ representa un número binario arbitrario
$$ x_2 = \pm (\alpha_n \, \alpha_{n-1} \, \cdots \, \alpha_{1} \,  \alpha_0  \, . \, \alpha_{-1} \, \alpha_{-2} \, \cdots) $$
donde $\alpha_i = 0$ o $\alpha_1 = 1$, la forma general de convertir números binarios a decimales es:
$$ x_{10} = \pm (\alpha_n \cdot 2^n + \alpha_{n-1} \cdot 2^{n-1} + \cdots + \alpha_{1} \cdot 2 + \alpha_0 \cdot 2^0 + \alpha_{-1} \cdot 2^{-1} + \alpha_{-2} \cdot 2^{-2}+ \cdots) $$
entonces, para el caso de $x_2=$ `110.0101`$_2$ tenemos:
$$ x_{10} = 1 \cdot 2^2 + 1 \cdot 2 + 0 \cdot 1 + 0 \cdot 2^{-1} + 1 \cdot 2^{-2} + 0 \cdot 2^{-3} + 1 \cdot 2^{-4} = \frac{101}{16} = 6.3125 $$

Realizar esta conversión en forma "manual" es muy tedioso, así que una buena opción es construir una función en Python que realice la tarea por nosotros.

La función `bin2dec()` recibe como argumento una cadena de ceros y unos (y eventualmente un punto decimal) que representa al número binario que queremos convertir. Si esta cadena tiene un punto decimal, divide la cadena en la "parte entera" (a la izquierda del punto) y en la "parte decimal" (a la derecha), y luego va multiplicando cada `1` que encuentra por la correspondiente potencia de 2. En caso que no haya un punto decimal en la cadena, considera que la misma representa un número entero. La función devuelve entonces el valor decimal del número binario.

In [5]:
def bin2dec(s):
    if not '.' in s:
        be, bd = s, ''
    else:
        be, bd = s.split('.')        
    d = 0
    for e in range(len(be)):
        d += int(be[-(e + 1)]) * 2**(e)
    for e in range(len(bd)):
        d += int(bd[e]) * 2**(-(e + 1))
    return(d)

Usamos entonces esta función recorriendo una lista con los números binarios que propone el ejercicio:

In [6]:
binarios = ['1011001', '110.0101', '0.01011']
for bin in binarios:
    print(f"{bin} -> {bin2dec(bin)}")

1011001 -> 89
110.0101 -> 6.3125
0.01011 -> 0.34375


## Ejercicio 4

La conversión de un [número hexadecimal](https://es.wikipedia.org/wiki/Sistema_hexadecimal) a uno decimal tiene el mismo procedimiento que la conversión desde un número binario a uno decimal, con la diferencia que la base ahora es 16, y en vez de ceros y unos ahora tenemos los diez dígitos usuales mas letras que expanden la base numérica. Esta base es:

$$ \{ 1, 2, 3, 4, 5, 6, 7, 8, 9, a, b, c, d, e, f, g, h\} $$

donde las letras corresponden a los valores decimales $a =10_{10}$, $b =11_{10}$, $c =12_{10}$, $d = 13_{10}$, $e =14_{10}$, $f =15_{10}$. En este contexto, $11_{\text{h}} = 16_{10}$.

Entonces, el número `2c.0b7`$_{\text{h}}$ en decimal es:
\begin{align}
2c.0b7_{\text{h}} &= 2 \cdot 16^1 + 12 \cdot 16^0 + 0 \cdot 16^{-1} + 11 \cdot 16^{-2} + 7 \cdot 16^{-3} \\
&= 32 + 12  + \frac{1}{16} + \frac{11}{256} + \frac{7}{4096} \\
&= \frac{180407}{4096} = 44.044677734375_{10}
\end{align}
Se puede notar que el sistema hexadecimal permite una representación más compacta de los números. En este ejemplo, con 5 *dígitos* hexadecimales podemos representar un número que en el sistema decimal requiere de 14.

## Ejercicio 5

Para convertir un número binario en representación de [punto flotante con precisión simple](https://es.wikipedia.org/wiki/Formato_en_coma_flotante_de_simple_precisi%C3%B3n) es necesario seguir los siguientes pasos:

1. Identificar los dígitos que representan el signo ($s$), el exponente ($E$) y la mantisa ($m$)
2. Convertir el exponente y la mantisa a números decimales
3. Determinar el exponente efectivo restando del obtenido previamente el *bias* 127 ($e = E - 127$)
4. Obtener el número decimal como $(-1)^s \, m \cdot 10^e$

Para el primer paso, sabemos que el bit de signo es el primero, los siguientes 8 bits corresponden al exponente y los restantes 23 bits representan la mantisa normalizada (esto es se asume que el primer bit antes del punto decimal es un 1 y no se almacena).

Entonces, descomponemos el número `01001010010000100011011001000000` en:

- `0` es el signo (al ser nulo representa un número **positivo**)
- `10010100` es el exponente
- `10000100011011001000000` es la mantisa (que lleva un `1` adelante que **no** se almacena).

Ahora aprovechamos la función `bin2dec()` que definimos en el ejercicio 3 para convertir el exponente y la mantisa al sistema decimal (adicionando el `1` aantes del punto, omitido en el formato):

In [7]:
bin2dec('10010100')

148

In [8]:
bin2dec('1.10000100011011001000000')

1.5172805786132812

Recordando que el *bias* es 127, el número buscado es:
$$ (-1)^0 \, 1.5172805786132812 \cdot 2^{148 - 127} = 1.5172805786132812 \cdot 2^{21} = 3181968 $$

En vez de repetir el cálculo *manualmente* para cada número del ejercicio, podemos construir una función en Python que lo haga por nosotros. La función `iee2dec()` recibe una cadena de caracteres que representa un número binario con el formato IEE-754 de precisión simple, identifica los bits correspondientes al signo, el exponente y la mantisa normalizada, y por medio de la función `bin2dec()` convierte el exponente y la mantisa al sistema decimal, y devuelve el número ingresado en forma decimal, y una cadena en la que se encuentran separados por espacios los tres campos del número binario:

In [9]:
def iee2dec(s):
    bs = int(s[0]) # bit de signo
    be = s[1:9]    # bits de exponente binario
    bmn = s[9:]    # bits de mantisa binaria normalizada
    e = bin2dec(be) - 127  # exponente decimal - bias
    m = '1.' + bmn  # 1 + mantisa normalizada
    md = bin2dec(m) # mantisa a decimal
    return (-1)**bs * md * 2**e, str(bs) + ' ' + be + ' ' + bmn

iee2dec('01001010010000100011011001000000')

(3181968.0, '0 10010100 10000100011011001000000')

In [10]:
lista_bin = ['01001010010000100011011001000000', 
             '11000100111100001110000000000000',
             '01000000010010010000111111011010']

for b in lista_bin:
    d, s = iee2dec(b)
    print(f"{s} -> {d}")

0 10010100 10000100011011001000000 -> 3181968.0
1 10001001 11100001110000000000000 -> -1927.0
0 10000000 10010010000111111011010 -> 3.141592502593994


## Ejercicio 6

En este ejercicio realizaremos operaciones aritméticas simples propagando el error según la regla para sumas/restas y productos/divisiones. En cada caso, si $a = \tilde{a} \pm \Delta a$ y $b = \tilde{b} \pm \Delta b$:

\begin{align}
a + b &= ( \tilde{a} + \tilde{b}) \pm (\Delta a + \Delta b) \\
a \cdot b &= (\tilde{a} \cdot \tilde{b}) \pm \left( \frac{\Delta a}{\tilde{a}} + \frac{\Delta{b}}{\tilde{b}} \right) (a \cdot b)
\end{align}
y del mismo modo para restas y divisiones. Para el último caso hemos aproximado el error relativo $\epsilon_r = \Delta a / a \approx \Delta a / \tilde{a}$, y si $c = a \cdot b$, el error absoluto de $c$ lo podemos obtener a partir de su error relativo:
$$ \epsilon_r = \frac{\Delta c}{c} \approx \frac{\Delta a}{\tilde{a}} + \frac{\Delta b}{\tilde{b}} \implies \Delta c = \epsilon_r * c $$

Entonces:

a) 
\begin{align}
(3.5 \pm 0.1) + (8.0 \pm 0.2) - (5.0 \pm 0.4) &= (3.5 + 8.0 - 5.0) \pm (0.1 +0.2 +0.4) \\
&= 6.5 \pm 0.7
\end{align}

b)
\begin{align}
(3.5 \pm 0.1) \cdot (8.0 \pm 0.2) &= (3.5 \cdot 8.0) \pm \left( \frac{0.1}{3.5} + \frac{0.2}{8.0} \right) \cdot (3.5 \cdot 8.0) \\
&= 28.0 \pm 0.05357 \cdot 28.0 \\
&= 28.0 \pm 1.5
\end{align}

c)
\begin{align}
(3.5 \pm 0.1) \times (8.0 \pm 0.2) / (5.0 \pm 0.4) &= \frac{3.5 \cdot 8.0}{5.0} \pm \left( \frac{0.1}{3.5} + \frac{0.2}{8.2} + \frac{0.4}{5.0} \right) \cdot \frac{3.5 \cdot 8.0}{5.0} \\
&= 5.6 \pm 0.133571 \cdot 5.6 \\
&= 5.60 \pm 0.75
\end{align}

## Ejercicio 7

Considerando el caso general de propagación de errores mediante la expansión en serie de Taylor de primer orden, y dado que en este ejercicio la función a evaluar tiene solo una variable independiente que podemos expresar como $\theta = \tilde{\theta} \pm \Delta \theta$, tenemos:

$$ f(\theta) \approx f(\tilde{\theta}) + \left. \frac{df}{d \theta} \right|_{\theta = \tilde{\theta}} \Delta \theta $$

Entonces:
$$ f(\tilde{\theta}) = \sin 125^{\circ} = 0.819152044288992 $$
$$ \left. \frac{d(\sin \theta)}{d \theta} \right|_{\theta = \tilde{\theta}} = \left. \cos \theta \right|_{\theta = \tilde{\theta}} = \cos 125^{\circ} = - 0.5735764363510458 $$
$$ \delta f \leq \left| \frac{df}{d \theta} \right|_{\theta = \tilde{\theta}} \, \Delta \theta = 0.5735764363510458 \cdot 0.034906585 = 0.02002 $$

En esta última expresión debemos expresar $\Delta \theta$ en radianes, ya que el error no debe tener unidades debido a que $\sin \theta$ es un número real sin unidades. El resultado, expresando el error con solo una cifra significativa, es:

$$ \sin (125 \pm 2)^{\circ} = 0.82 \pm 0.02 $$